src/oic/oic/provider.py

```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
+
+    def do_back_channel_logout(
+        self, cinfo: dict, sub: str, sid: str
+    ) -> Optional[Tuple[str, str]]:
+        """
+        Prepare information to be used to do a back-channel logout.
+
+        :param cinfo: Client information
+        :param sub: Subject identifier
+        :param sid: The Issuer ID
+        :return: Tuple with logout URI and signed logout token
+        """
+        try:
+            back_channel_logout_uri = cinfo["backchannel_logout_uri"]
+        except KeyError:
+            return None
+
+        # always include sub and sid so I don't check for
+        # backchannel_logout_session_required
+
+        payload = {
+            "sub": sub,
+            "sid": sid,
+            "events": {BACK_CHANNEL_LOGOUT_EVENT: {}},
+            "jti": uuid.uuid4().hex,
+        }
+
+        try:
+            alg = cinfo["id_token_signed_response_alg"]
+        except KeyError:
+            alg = self.capabilities["id_token_signing_alg_values_supported"][0]
+
+        _jws = JWT(self.keyjar, iss=self.name, lifetime=86400, sign_alg=alg)
+        sjwt = _jws.pack(aud=cinfo["client_id"], **payload)
+
+        return back_channel_logout_uri, sjwt
+
+    def clean_sessions(self, usids: List[str]):
+        """
+        Remove Session IDs from the session DB.
+
+        :param usids: List of session IDs
+        """
+        _sdb = self.sdb
+        # Clean out all sessions
+        for sid in usids:
+            del _sdb[sid]
+
+    def logout_all_clients(self, sid: str) -> Dict:
+        """
+        Collect information necessary to logout one user from all clients he/she has been using.
+
+        :param sid: Session ID
+        :return: Dictionary with logout information
+        """
+        _sdb = self.sdb
+
+        uid = _sdb.get_uid_by_sid(sid)
+        _client_sid = {}
+        # Find all the session IDs this user has gotten
+        usids = session_get(_sdb, "uid", uid)
+        # Find all RPs this user has logged it from
+        for usid in usids:
+            _client_sid[_sdb[usid]["client_id"]] = usid
+
+        # Front-/Backchannel logout ?
+        _cdb = self.cdb
+        _iss = self.name
+        bc_logouts = {}
+        fc_iframes = {}
+        for _cid, _csid in _client_sid.items():
+            if "backchannel_logout_uri" in _cdb[_cid]:
+                _sub = _sdb[_csid]["sub"]
+                _spec = self.do_back_channel_logout(_cdb[_cid], _sub, _csid)
+                if _spec:
+                    bc_logouts[_cid] = _spec
+            elif "frontchannel_logout_uri" in _cdb[_cid]:
+                # Construct an IFrame
+                _iframe = do_front_channel_logout_iframe(_cdb[_cid], _iss, _csid)
+                if _iframe:
+                    fc_iframes[_cid] = _iframe
+
+        # Clean out all sessions
+        self.clean_sessions(usids)
+
+        return {"blu": bc_logouts, "flu": fc_iframes}
+
+    def logout_from_client(self, sid: str, client_id: str) -> Dict:
```

Comment:
This method would benefit for more descriptive variable names... It is sort of unreadable...

Reply:
OK